File: src/main/java/com/meti/server/route/AssetRoute.java
package com.meti.server.route;

import com.meti.server.context.Request;
import com.meti.server.response.InlineResponse;
import com.meti.server.response.Response;
import com.meti.server.response.ResponseCodes;
import com.meti.server.response.ResponseTypes;

import java.io.IOException;
import java.io.InputStream;
import java.nio.file.Path;

import static java.nio.file.Files.exists;
import static java.nio.file.Files.newInputStream;

public class AssetRoute implements Route {
    private final Path internalPath;
    private final String virtualPath;

    public AssetRoute(Path internalPath, String virtualPath) {
        this.internalPath = internalPath;
        this.virtualPath = virtualPath;
    }

    @Override
    public boolean canProcess(Request request) {
        return exists(buildPath(request));
    }

    @Override
    public Response process(Request request) {
        try {
            return processExceptionally(request);
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    private Response processExceptionally(Request request) throws IOException {
        Path asset = buildPath(request);
        byte[] bytes = readAllBytes(asset);
        return new InlineResponse(ResponseCodes.OK, ResponseTypes.get(asset), bytes);
    }

    private byte[] readAllBytes(Path asset) throws IOException {
        byte[] bytes;
        try (InputStream inputStream = newInputStream(asset)) {
            bytes = inputStream.readAllBytes();
        }
        return bytes;
    }

    private Path buildPath(Request request) {
        String path = request.getPath();
        String child = path.replace(virtualPath + "/", "");
        return internalPath.resolve(child);
    }
}

File: src/main/scala/com/meti/route/AssetRoute.scala
package com.meti.route

import java.io.IOException
import java.nio.file.{Files, Path, Paths}

import com.meti.Request
import com.meti.response._

case class AssetRoute(internalPath: Path, webPath: String) extends Route {
  override def canProcess(request: Request): Boolean = Files.exists(buildPath(request))

  override def process(request: Request): Response = {
    try {
      val path = buildPath(request)
      val data = Files.readAllBytes(path)
      val nanoType = NanoType(buildPath(request).toUri)
      InlineResponse(DefaultCode.OK, nanoType, data)
    } catch {
      case e: IOException => StringResponse(DefaultCode.INTERNAL_SERVER_ERROR, PlainType, e.getLocalizedMessage)
    }
  }

  private def buildPath(request: Request): Path = internalPath.resolve(buildArgsPath(request))

  private def buildArgsPath(request: Request) = request.getPath()
    .replace(webPath + "/", "")
    .split("/")
    .map((value: String) => Paths.get(value))
    .reduce((path0: Path, path1: Path) => path0.resolve(path1))
}
